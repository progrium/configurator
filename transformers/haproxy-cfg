#!/usr/bin/ruby
require "rubygems"
require "json"

$proxy_keywords = ["acl", "appsession", "backlog", "balance", "bind", "bind-process", "block", "capture", 
	"clitimeout", "contimeout", "cookie", "default-server", "default_backend", "description", "disabled", 
	"dispatch", "enabled", "errorfile", "errorloc", "errorloc302", "errorloc303", "force-persist", "fullconn", 
	"grace", "hash-type", "http-check", "http-request", "id", "ignore-persist", "log", "maxconn", "mode", 
	"monitor fail", "monitor-net", "monitor-uri", "option", "persist rdp-cookie", "rate-limit sessions", 
	"redirect", "redisp", "redispatch", "reqadd", "reqallow", "reqdel", "reqdeny", "reqiallow", "reqidel", 
	"reqideny", "reqipass", "reqirep", "reqisetbe", "reqitarpit", "reqpass", "reqrep", "reqsetbe", "reqtarpit", 
	"retries", "rspadd", "rspdel", "rspdeny", "rspidel", "rspideny", "rspirep", "rsprep", "server", "source", 
	"srvtimeout", "stats", "stick", "stick-table", "tcp-request", "timeout", "transparent", "use_backend"]
$keyword_sections = {
	"global" 	=> ["chroot", "daemon", "gid", "group", "log", "log-send-hostname", "nbproc", "pidfile", 
		"uid", "ulimit-n", "user", "stats", "node", "description", "maxconn", "maxpipes", "noepoll", "nokqueue", 
		"nopoll", "nosepoll", "nosplice", "spread-checks", "tune.bufsize", "tune.chksize", "tune.maxaccept", 
		"tune.maxpollevents", "tune.maxrewrite", "tune.rcvbuf.client", "tune.rcvbuf.server", "tune.sndbuf.client", 
		"tune.sndbuf.server", "debug", "quiet"],
	"userlist"	=> ["group", "user"],
	"defaults" 	=> $proxy_keywords,
	"listen" 	=> $proxy_keywords,
	"frontend" 	=> $proxy_keywords,
	"backend" 	=> $proxy_keywords
}
$named_sections = ["listen", "frontend", "backend"] # intentionally not including "defaults"

def strip_extra_spaces(str)
	str.split("\n").collect{|s| s.gsub(/\s+/, " ").strip }.join "\n"
end

def arg(obj)
	case obj
	when Array
		obj.collect{|a| arg(a) }.join " "
	when Hash
		obj.collect{|k,v| [k, arg(v)].join(" ") }.join " "
	when TrueClass, NilClass, FalseClass
		""
	when Fixnum, Float
		obj.to_s
	when String
		obj
	else
		raise "unexpected value for arg #{obj}"
	end
end

def keyword(name, args)
	if args.is_a? Hash
		args.collect do |k,v|
			if v == false
				keyword("no #{name} #{k}", v)
			else
				keyword("#{name} #{k}", v)
			end
		end .compact.join "\n"
	elsif args.is_a?(Array) and args.first.is_a?(Array)
		args.collect do |line| 
			arg([name] + line)
		end .join "\n"
	else
		"#{name} #{arg(args)}"
	end
end

def section(name, body)
	section_keyword = name.split(" ").first
	raise "unexpected section #{section_keyword}" \
		unless $keyword_sections.keys.include? section_keyword
	if $named_sections.include? name
		if body.is_a? Hash
			body.collect{|k,v| section("#{name} #{k}", v) }.join
		else
			raise "named section #{name} expecting object value"
		end
	else
		lines = case body
		when Hash
			body.collect do |kw, args|
				raise "keyword not allowed in #{section_keyword}: #{kw}" \
					unless $keyword_sections[section_keyword].include? kw
				strip_extra_spaces(keyword(kw, args))
			end
		when Array
			body.collect do |line|
				raise "keyword not allowed in #{section_keyword}: #{line.first}" \
					unless $keyword_sections[section_keyword].include? line.first
				strip_extra_spaces(args(line))
			end
		end .join "\n"
		["#{name}", lines.gsub(/^/, "  "), "\n"].join "\n"
	end
end

def transform(obj)
	raise "non object as root" unless obj.is_a? Hash
	obj.collect do |name, body|
		section(name, body)
	end .join
end

puts transform(JSON.parse(STDIN.read))
