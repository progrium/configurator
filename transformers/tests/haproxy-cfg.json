{ 
  "global": {
    "daemon": true,
    "chroot": "/var/lib/haproxy",
    "user": "haproxy",
    "group": "haproxy",
    "maxconn": 4096,
    "log": [
      ["logs.papertrailapp.com:45187", "local0"],
      ["logs.papertrailapp.com:45187", "local1", "emerg", "notice"]
    ]
  },
  "defaults": {
    "log": "global",
    "mode": "http",
    "option": {
      "httplog": true,
      "dontlognull": true,
      "http-tunnel": true
    },
    "timeout": {
      "connect": 5000,
      "client": "50s",
      "server": "1m"
    },
    "errorfile": {
      "400": "/etc/haproxy/errors/400.http",
      "403": "/etc/haproxy/errors/403.http"
    }
  },
  "listen": {
    "stats": {
      "bind": "0.0.0.0:8080",
      "mode": "http",
      "stats": {
        "enable": true,
        "uri": "/",
        "refresh": "5s"
      }
    }
  },
  "frontend": {
    "all": {
      "bind": [
        ["192.168.0.5:80"],
        ["*:8040"]
      ],
      "default_backend": "appin",
      "mode": "http",
      "acl": {
        "is_websocket": ["hdr(Upgrade)", "-i", "WebSocket"]
      },
      "use_backend": ["websocket", "if", "is_websocket"],
      "option": {
        "forwardfor": ["except", "127.0.0.1"],
        "httplog": ["clf"]
      }
    }
  },
  "backend": {
    "appin": {
      "balance": ["url_param", "userid"],
      "timeout": {
        "queue": 5000,
        "server": 50000
      },
      "option": {
        "forwardfor": ["header", "X-Client"],
        "http-tunnel": false
      },
      "server": {
        "socket0": ["localhost:6000", "maxconn", 500, "check"],
        "socket1": ["localhost:6001", "maxconn", 500, "check"]
      }
    },
    "websocket": {
      "balance": ["leastconn"],
      "timeout": {
        "queue": 5000,
        "server": 86400000
      },
      "option": {
        "httpchk": ["GET", "/ws-healthcheck", "HTTP/1.0"]
      },
      "server": {
        "socket10": ["localhost:6010", {"maxconn": 10000, "check": true, "inter": "20s"}],
        "socket11": ["localhost:6011", {"maxconn": 10000, "check": true, "backup": true}],
        "socket12": ["localhost:6012", {"maxconn": 10000, "check": true, "inter": 5000}],
        "socket13": ["localhost:6013", {"maxconn": 10000, "check": true, "source": ["192.168.0.5:80", {"interface": "lo"}]}],
        "socket14": ["localhost:6014", {"maxconn": 10000, "check": true, "source": ["192.168.0.5:80", {"interface": "eth0"}]}]
      }
    }
  }
}
