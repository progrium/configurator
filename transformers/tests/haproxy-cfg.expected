global
  daemon
  chroot /var/lib/haproxy
  user haproxy
  group haproxy
  maxconn 4096
  log logs.papertrailapp.com:45187 local0
  log logs.papertrailapp.com:45187 local1 emerg notice

defaults
  log global
  mode http
  option httplog
  option dontlognull
  option http-tunnel
  timeout connect 5000
  timeout client 50s
  timeout server 1m
  errorfile 400 /etc/haproxy/errors/400.http
  errorfile 403 /etc/haproxy/errors/403.http

listen stats
  bind 0.0.0.0:8080
  mode http
  stats enable
  stats uri /
  stats refresh 5s

frontend all
  bind 192.168.0.5:80
  bind *:8040
  default_backend appin
  mode http
  acl is_websocket hdr(Upgrade) -i WebSocket
  use_backend websocket if is_websocket
  option forwardfor except 127.0.0.1
  option httplog clf

backend appin
  balance url_param userid
  timeout queue 5000
  timeout server 50000
  option forwardfor header X-Client
  no option http-tunnel
  server socket0 localhost:6000 maxconn 500 check
  server socket1 localhost:6001 maxconn 500 check

backend websocket
  balance leastconn
  timeout queue 5000
  timeout server 86400000
  option httpchk GET /ws-healthcheck HTTP/1.0
  server socket10 localhost:6010 maxconn 10000 check inter 20s
  server socket11 localhost:6011 maxconn 10000 check backup
  server socket12 localhost:6012 maxconn 10000 check inter 5000
  server socket13 localhost:6013 maxconn 10000 check source 192.168.0.5:80 interface lo
  server socket14 localhost:6014 maxconn 10000 check source 192.168.0.5:80 interface eth0
